jupyter_client/launcher.py

```diff
@@ -40,9 +41,103 @@ def launch_kernel(cmd, stdin=None, stdout=None, stderr=None, env=None,
     Returns
     -------
 
-    Popen instance for the kernel subprocess
+    subprocess.Popen instance for the kernel subprocess
+    """
+
+    kwargs, interrupt_event = prepare_process_args(stdin=stdin, stdout=stdout, stderr=stderr, env=env,
+                  independent=independent, cwd=cwd, **kw)
+
+    try:
+        # Allow to use ~/ in the command or its arguments
+        cmd = list(map(os.path.expanduser, cmd))
+
+        proc = Popen(cmd, **kwargs)
+    except Exception as exc:
+        msg = (
+            "Failed to run command:\n{}\n"
+            "    PATH={!r}\n"
+            "    with kwargs:\n{!r}\n"
+        )
+        # exclude environment variables,
+        # which may contain access tokens and the like.
+        without_env = {key:value for key, value in kwargs.items() if key != 'env'}
```

Comment:
Might be good to do a try / finally block around these so if there's message formatting or logging errors it still raises the original exception

Reply:
Added additional try/except block in https://github.com/jupyter/jupyter_client/pull/612/commits/e6e381ecea109f1564a27c15ac06a2427917f9cc.